File: nucleo/impacto.py
"""
Cálculo e seleção de impactos em p.p. para a nota de inflação.

A fórmula é determinística: impacto (p.p.) = peso_mensal × variação / 100.
A propriedade ItemInflacao.impacto já aplica a fórmula; este módulo fornece
as funções de filtragem usadas pelo montador.
"""
from fontes.modelo import ItemInflacao


def calcular_impacto(variacao: float, peso: float) -> float:
    """impacto em p.p. = peso × variação / 100, arredondado a 4 casas."""
    return round(peso * variacao / 100, 4)


def grupos_relevantes(
    grupos: list[ItemInflacao],
    top_n: int = 3,
    threshold: float = 0.05,
) -> list[ItemInflacao]:
    """
    Retorna até top_n grupos com maior impacto absoluto acima de threshold p.p.
    A entrada deve estar ordenada por impacto desc (padrão de ibge.buscar_resultado).
    """
    return [g for g in grupos[:top_n] if abs(g.impacto) >= threshold]


def top_subitem(
    subitens: list[ItemInflacao],
    nivel: int = 4,
) -> ItemInflacao | None:
    """
    Retorna o subitem de maior impacto positivo no nível solicitado, ou None se vazio.
    nivel=4 → subitem próprio; nivel=2 → subgrupo (fallback se não houver nivel 4).
    """
    candidatos = [s for s in subitens if s.nivel == nivel and s.impacto > 0]
    return candidatos[0] if candidatos else None


def grupos_queda(
    grupos: list[ItemInflacao],
    top_n: int = 1,
    threshold: float = 0.05,
) -> list[ItemInflacao]:
    """
    Retorna até top_n grupos com maior deflação (impacto <= -threshold p.p.),
    ordenados do mais negativo ao menos negativo.
    A entrada deve estar ordenada por impacto desc (padrão de ibge.buscar_resultado).
    """
    queda = [g for g in grupos if g.impacto <= -threshold]
    queda.sort(key=lambda x: x.impacto)  # mais negativo primeiro
    return queda[:top_n]


def top_subitem_queda(
    subitens: list[ItemInflacao],
    nivel: int = 4,
    threshold: float = 0.05,
) -> ItemInflacao | None:
    """
    Retorna o subitem de maior deflação (impacto <= -threshold) no nível solicitado.
    """
    candidatos = [s for s in subitens if s.nivel == nivel and s.impacto <= -threshold]
    if not candidatos:
        return None
    return min(candidatos, key=lambda x: x.impacto)


def calcular_difusao(
    subitens: list[ItemInflacao],
    nivel_min: int = 4,
) -> float | None:
    """
    Parcela de subitens (nivel >= nivel_min) com variação positiva no mês, em %.
    Usado pelo montador e para difusão do IPCA-15 (sem série SGS própria).
    """
    candidatos = [s for s in subitens if s.nivel >= nivel_min]
    if not candidatos:
        return None
    positivos = sum(1 for s in candidatos if s.variacao > 0)
    return round(positivos / len(candidatos) * 100, 2)

File: site/js/impacto.js
/* Seleção de impactos — par de nucleo/impacto.py.
 *
 * Nota sobre ordenação: Array.prototype.sort é estável desde o ES2019, assim
 * como o sort do Python. Empates de impacto preservam a ordem de entrada nos
 * dois, então a seleção de destaques é a mesma — desde que a ordem original
 * venha da resposta da API, que ibge.js preserva com Map.
 */
import { arredondar } from "./numeros.js";

/** impacto em p.p. = peso × variação / 100, arredondado a 4 casas. */
export function calcularImpacto(variacao, peso) {
  return arredondar((peso * variacao) / 100, 4);
}

/** Até topN grupos com impacto absoluto acima de threshold p.p. */
export function gruposRelevantes(grupos, topN = 3, threshold = 0.05) {
  return grupos.slice(0, topN).filter((g) => Math.abs(g.impacto) >= threshold);
}

/** Item de maior impacto positivo no nível pedido, ou null. Entrada já ordenada desc. */
export function topSubitem(subitens, nivel = 4) {
  return subitens.find((s) => s.nivel === nivel && s.impacto > 0) ?? null;
}

/** Até topN grupos com maior deflação, do mais negativo ao menos negativo. */
export function gruposQueda(grupos, topN = 1, threshold = 0.05) {
  return grupos
    .filter((g) => g.impacto <= -threshold)
    .sort((a, b) => a.impacto - b.impacto)
    .slice(0, topN);
}

/** Item de maior deflação no nível pedido, ou null. */
export function topSubitemQueda(subitens, nivel = 4, threshold = 0.05) {
  const candidatos = subitens.filter(
    (s) => s.nivel === nivel && s.impacto <= -threshold,
  );
  if (candidatos.length === 0) return null;
  // Comparação estrita: empate devolve o primeiro, como o min() do Python.
  return candidatos.reduce((menor, s) => (s.impacto < menor.impacto ? s : menor));
}

/** Parcela de subitens (nivel >= nivelMin) com variação positiva, em %. */
export function calcularDifusao(subitens, nivelMin = 4) {
  const candidatos = subitens.filter((s) => s.nivel >= nivelMin);
  if (candidatos.length === 0) return null;
  const positivos = candidatos.filter((s) => s.variacao > 0).length;
  return arredondar((positivos / candidatos.length) * 100, 2);
}
